File: python/student_manager.py
import json
import os

# File name for storing student records (in the same folder as this script)
DATA_FILE = "student_records.json"

def load_student_records():
    """Load student records from a JSON file. Returns a dictionary."""
    if not os.path.exists(DATA_FILE):
        return {}  # Return empty if file doesn't exist
    with open(DATA_FILE, "r") as file:
        return json.load(file)

def save_student_records(records):
    """Save the given records dictionary to a JSON file."""
    with open(DATA_FILE, "w") as file:
        json.dump(records, file, indent=4)

def add_student_record(student_id, name, grades):
    """Add a new student record if one with the same ID doesn't already exist."""
    records = load_student_records()
    if student_id in records:
        print("Student record already exists!")
        return
    records[student_id] = {
        "name": name,
        "grades": grades
    }
    save_student_records(records)
    print("Student added successfully!")

def get_student_record(student_id):
    """Retrieve a student record by student ID."""
    records = load_student_records()
    return records.get(student_id)

def update_student_record(student_id, new_grades):
    """Update the grades of an existing student record."""
    records = load_student_records()
    if student_id not in records:
        print("Student not found!")
        return
    records[student_id]["grades"] = new_grades
    save_student_records(records)
    print("Student record updated successfully!")

def delete_student_record(student_id):
    """Delete a student record."""
    records = load_student_records()
    if student_id in records:
        del records[student_id]
        save_student_records(records)
        print("Student record deleted!")
    else:
        print("Student record not found!")

File: python/kotlin-app/app/src/main/kotlin/StudentManager.kt
import kotlinx.serialization.*
import kotlinx.serialization.json.*
import java.io.File


object StudentManager {
    private const val DATA_FILE = "src/main/resources/student_records.json"

    private fun loadRecords(): MutableMap<String, StudentRecord> {
        val file = File(DATA_FILE)
        return if (file.exists()) {
            Json.decodeFromString(file.readText())
        } else mutableMapOf()
    }

    private fun saveRecords(records: Map<String, StudentRecord>) {
        File(DATA_FILE).writeText(Json.encodeToString(records))
    }

    fun addStudentRecord(id: String, name: String, grades: List<Double>) {
        val records = loadRecords().toMutableMap()
        if (records.containsKey(id)) {
            println("Student record already exists!")
            return
        }
        records[id] = StudentRecord(name, grades)
        saveRecords(records)
        println("Student added successfully!")
    }

    fun getStudentRecord(id: String): StudentRecord? = loadRecords()[id]

    fun updateStudentRecord(id: String, newGrades: List<Double>) {
        val records = loadRecords().toMutableMap()
        records[id]?.let {
            records[id] = it.copy(grades = newGrades)
            saveRecords(records)
            println("Student record updated successfully!")
        } ?: println("Student not found!")
    }

    fun deleteStudentRecord(id: String) {
        val records = loadRecords().toMutableMap()
        if (records.remove(id) != null) {
            saveRecords(records)
            println("Student record deleted!")
        } else {
            println("Student record not found!")
        }
    }

    fun loadAllRecords(): Map<String, StudentRecord> = loadRecords()
}